juju/controller.py

```diff
@@ -33,11 +33,28 @@ def __init__(self, loop=None, max_frame_size=None, bakery_client=None):
     def loop(self):
         return self._connector.loop
 
-    async def connect(self, controller_name=None):
-        """Connect to a Juju controller by name. If controller_name
-        is empty, connect to the current controller.
+    async def connect(self, controller_name=None, **kwargs):
+        """Connect to a Juju controller.
+
+        If any arguments are specified other than model_name,
+        then controller_name must be None and an explicit
+        connection will be made using Connection.connect
+        using those parameters (the 'uuid' parameter must
+        be absent or None).
+
+        Otherwise, if controller_name is None, connect to the
+        current controller.
+
+        Otherwise, controller_name must specify the name
+        of a known controller.
         """
-        await self._connector.connect_controller(controller_name)
+        await self.disconnect()
+        if len(kwargs) == 0:
+            await self._connector.connect_controller(controller_name)
```

Comment:
Shouldn't we exit the function here? Otherwise iirc we end up calling `self._connector.connect(**kwargs)` again below.

Reply:
I don't understand this comment. Each call to _connector.connect is a separate arm of the if statement. I don't see how both could happen.